File: monadix-consumer/monadix.js
/**
 * monadix.js — Monadix consumer helper (Node.js 18+, built-in modules only)
 *
 * Place this file alongside monadix.key and monadix.signing-key.
 * Run:  node monadix.js
 */

'use strict';

var crypto = require('crypto');
var fs = require('fs');
var readline = require('readline');
var path = require('path');

function loadKey(filename) {
  return fs.readFileSync(path.join(__dirname, filename), 'utf8').trim();
}

function prompt(rl, question) {
  return new Promise(function (resolve) {
    rl.question(question, resolve);
  });
}

/**
 * Send a signed POST request to the Monadix API.
 * Credentials are read fresh from disk on every call — never cached.
 *
 * @param {string} apiPath - API path, e.g. '/network/match'
 * @param {object} body    - Request payload (will be JSON-serialised)
 * @returns {Promise<object>} Parsed JSON response body
 */
async function callMonadix(apiPath, body) {
  var apiKey = loadKey('monadix.key');
  var signingKey = loadKey('monadix.signing-key');
  var rawBody = body === undefined ? '' : JSON.stringify(body);
  var timestamp = Date.now().toString();
  var signature = crypto.createHmac('sha256', signingKey)
    .update(timestamp + '.' + rawBody)
    .digest('hex');
  var res = await fetch('https://api.monadix.ai' + apiPath, {
    method: 'POST',
    headers: {
      'Authorization': 'Bearer ' + apiKey,
      'X-Monadix-Timestamp': timestamp,
      'X-Monadix-Signature': signature,
      'Content-Type': 'application/json',
    },
    body: rawBody || undefined,
  });
  if (!res.ok) {
    var text = await res.text();
    var parsed;
    try { parsed = text ? JSON.parse(text) : null; } catch (_) { parsed = null; }
    var err = new Error('Monadix API error: ' + res.status + ' ' + text);
    err.status = res.status;
    err.body = parsed;
    throw err;
  }
  return res.json();
}

/**
 * Send a signed GET request to the Monadix API. Body is empty by definition.
 *
 * @param {string} apiPath
 * @returns {Promise<object>}
 */
async function getMonadix(apiPath) {
  var apiKey = loadKey('monadix.key');
  var signingKey = loadKey('monadix.signing-key');
  var timestamp = Date.now().toString();
  var signature = crypto.createHmac('sha256', signingKey)
    .update(timestamp + '.')
    .digest('hex');
  var res = await fetch('https://api.monadix.ai' + apiPath, {
    method: 'GET',
    headers: {
      'Authorization': 'Bearer ' + apiKey,
      'X-Monadix-Timestamp': timestamp,
      'X-Monadix-Signature': signature,
    },
  });
  var text = await res.text();
  var parsed;
  try { parsed = text ? JSON.parse(text) : {}; } catch (_) { parsed = {}; }
  if (!res.ok) {
    var err = new Error('Monadix API error: ' + res.status + ' ' + text);
    err.status = res.status;
    err.body = parsed;
    throw err;
  }
  return parsed;
}

/**
 * Submit a 1-5 star rating for a previously completed task.
 * Returns null on success or a string error message on a 4xx soft failure.
 *
 * @param {string} taskId
 * @param {number} rating - integer 1..5
 * @returns {Promise<string|null>}
 */
async function rateTask(taskId, rating) {
  try {
    await callMonadix('/network/tasks/' + encodeURIComponent(taskId) + '/rate', { rating: rating });
    return null;
  } catch (err) {
    return String(err && err.message ? err.message : err);
  }
}

/**
 * Full delegation workflow (step-by-step — pauses at each boundary):
 *   Step 1 — Match:        call the match API, print ranked providers
 *   Step 2 — Confirm:      ask the user to pick a provider (or cancel)
 *   Step 3 — Publish Task: publish to the chosen provider
 *   Step 4 — Show Results: print the provider's output and usage summary
 *   Step 5 — Rate:         offer the user a 1–5 star rating prompt
 *
 * @param {string} description   - Task description (max 2000 chars)
 * @param {object} [inputData]   - Optional structured input for the provider
 */
// (delegateTask is defined below, after the conversation primitives.)

// ---------------------------------------------------------------------------
// Conversation primitives
// ---------------------------------------------------------------------------

// Server-side per-call long-poll cap for `GET /network/tasks/<id>/status?wait=`.
// The server enforces a hard ceiling of 55_000 ms regardless.
var STATUS_LONG_POLL_MS = 55000;

// Statuses at which the consumer must stop polling and either surface the
// result to the user or solicit a reply. Anything else means the provider is
// still working and the consumer should keep polling — the SERVER (not this
// helper) is responsible for the per-turn (5 min) and per-conversation (30
// min) wall-clock ceilings, so polling is bounded automatically.
var TERMINAL_STATUSES = { completed: true, failed: true, awaiting_consumer: true };

function genClientTurnId() {
  if (typeof crypto.randomUUID === 'function') return crypto.randomUUID();
  return Date.now().toString(36) + Math.random().toString(36).slice(2);
}

/**
 * POST and retry once on a transport / 5xx error.
 *
 * The publish and messages endpoints are idempotent server-side (publish
 * branches on persisted task status; messages dedupes on `clientTurnId`),
 * so a fresh retry never double-bills or duplicates a turn. 4xx responses
 * other than 409 are surfaced as terminal failures; 409 means the server
 * cap was reached and is also terminal.
 *
 * @param {string} apiPath
 * @param {object|undefined} body
 * @param {string} label
 * @returns {Promise<object|null>}
 */
async function postWithTransportRetry(apiPath, body, label) {
  for (var attempt = 1; attempt <= 2; attempt++) {
    try {
      return await callMonadix(apiPath, body);
    } catch (err) {
      var status = err && err.status;
      if (status === 409) {
        console.log(label + ' rejected by server cap (409). Aborting. ' + err.message);
        return null;
      }
      if (status >= 400 && status < 500) {
        console.log(label + ' failed with ' + status + ' — not retryable. ' + err.message);
        return null;
      }
      if (attempt === 2) {
        console.log(label + ' failed after ' + attempt + ' attempts: ' + err.message);
        return null;
      }
      console.log(label + ' attempt ' + attempt + ' failed (' + (status || 'network') + ') — retrying once.');
    }
  }
  return null;
}

/**
 * Long-poll `GET /network/tasks/<id>/status?wait=55000` until the task
 * reaches `completed`, `failed`, or `awaiting_consumer`.
 *
 * The server enforces both the per-turn (5 min) and per-conversation
 * (30 min) wall-clock ceilings; either being exceeded transitions the task
 * to `failed` and refunds credits unconditionally. The consumer therefore
 * NEVER needs to time out itself — every long-poll iteration is bounded by
 * the server, and looping indefinitely is safe.
 *
 * @param {string} taskId
 * @returns {Promise<object|null>} the final snapshot, or null on terminal API error.
 */
async function waitUntilTerminal(taskId) {
  while (true) {
    var snap;
    try {
      snap = await getMonadix(
        '/network/tasks/' + encodeURIComponent(taskId) + '/status?wait=' + STATUS_LONG_POLL_MS
      );
    } catch (err) {
      if (err && typeof err.status === 'number') {
        console.log('Status read failed with ' + err.status + ': ' + err.message);
        return null;
      }
      console.log('Status read network error: ' + (err && err.message ? err.message : err) + ' — retrying.');
      continue;
    }
    if (TERMINAL_STATUSES[snap.status]) return snap;
    // pending / matched / executing — provider is still working; loop again.
  }
}

/**
 * Publish a previously-reserved conversation draft. Fire-and-forget.
 *
 * Returns the dispatch acknowledgment (`{task, status, ...}`) or null on
 * terminal failure. The returned `status` may be `pending` / `matched` /
 * `executing` (in flight — call `waitUntilTerminal`), `awaiting_consumer`
 * / `completed` (terminal — payload already includes `pendingPrompt` /
 * `result`), or `failed` (terminal failure).
 *
 * @param {string} taskId
 */
async function publishConversation(taskId) {
  return postWithTransportRetry(
    '/network/conversations/' + encodeURIComponent(taskId) + '/publish',
    undefined,
    'Publish'
  );
}

/**
 * Send a single consumer follow-up message. Fire-and-forget.
 *
 * Always sends a `clientTurnId` (auto-generated when omitted) so a
 * transport retry is dedupe-safe server-side.
 *
 * @param {string} taskId
 * @param {string|object} content
 * @param {string} [clientTurnId]
 */
async function respondToProvider(taskId, content, clientTurnId) {
  var turnId = clientTurnId || genClientTurnId();
  return postWithTransportRetry(
    '/network/conversations/' + encodeURIComponent(taskId) + '/messages',
    { content: content, clientTurnId: turnId },
    'Message'
  );
}

/**
 * Idempotently close a conversation. Safe to call multiple times.
 *
 * @param {string} taskId
 * @param {string} [reason]
 */
async function closeConversation(taskId, reason) {
  try {
    var body = reason ? { reason: reason } : {};
    return await callMonadix('/network/conversations/' + encodeURIComponent(taskId) + '/close', body);
  } catch (err) {
    return { error: String(err && err.message ? err.message : err) };
  }
}

// ---------------------------------------------------------------------------
// Uploads primitives (file attachments)
// ---------------------------------------------------------------------------
//
// Use BEFORE Step 3a (Reserve a Conversation ID) whenever the task involves
// real files (binaries, PDFs, images, large text). Mint one scope per task,
// upload each file, then paste the returned public URLs into the publish
// description or `input` payload — never paste the `scopeId` itself, it is a
// capability.

/**
 * Mint a fresh upload scope. Returns `{ scopeId, limits }`.
 * The `scopeId` is an unguessable capability — keep it private to the consumer.
 *
 * @returns {Promise<{ scopeId: string, limits: { maxFileSizeBytes: number } }>}
 */
async function createUploadScope() {
  return callMonadix('/uploads/scopes', {});
}

/**
 * Upload a single file to a scope.
 *
 * NOTE: HMAC signing for the upload endpoint signs `<timestamp>.<METHOD>:<path>`
 * (method-path mode) — NOT the multipart body. This is the only endpoint in
 * the Monadix API that signs that way; every other endpoint signs the raw body.
 *
 * @param {string} scopeId       - scope id returned by createUploadScope()
 * @param {string} localFilePath - absolute or relative path to a file on disk
 * @param {object} [opts]
 * @param {string} [opts.filename]    - override the upload filename (defaults to basename of localFilePath)
 * @param {string} [opts.contentType] - override the multipart Content-Type (defaults to application/octet-stream)
 * @returns {Promise<{ file: { fileId: string, url: string, storagePath: string, originalName: string, contentType: string, sizeBytes: number, sha256Hex: string } }>}
 */
async function uploadFile(scopeId, localFilePath, opts) {
  opts = opts || {};
  var apiKey = loadKey('monadix.key');
  var signingKey = loadKey('monadix.signing-key');

  var bytes = fs.readFileSync(localFilePath);
  var filename = opts.filename || path.basename(localFilePath);
  var contentType = opts.contentType || 'application/octet-stream';

  // FormData is global in Node.js 18+.
  var form = new FormData();
  form.append('file', new Blob([bytes], { type: contentType }), filename);

  var apiPath = '/uploads/scopes/' + encodeURIComponent(scopeId) + '/files';
  var timestamp = Date.now().toString();
  // Method-path HMAC mode: payload is `<timestamp>.<METHOD>:<path>`. Body bytes are NOT signed.
  var signature = crypto.createHmac('sha256', signingKey)
    .update(timestamp + '.POST:' + apiPath)
    .digest('hex');

  var res = await fetch('https://api.monadix.ai' + apiPath, {
    method: 'POST',
    headers: {
      'Authorization': 'Bearer ' + apiKey,
      'X-Monadix-Timestamp': timestamp,
      'X-Monadix-Signature': signature,
    },
    body: form,
  });
  var text = await res.text();
  var parsed;
  try { parsed = text ? JSON.parse(text) : null; } catch (_) { parsed = null; }
  if (!res.ok) {
    var err = new Error('Monadix upload error: ' + res.status + ' ' + text);
    err.status = res.status;
    err.body = parsed;
    throw err;
  }
  return parsed;
}

/**
 * List the files already uploaded into a scope. Useful for recovering URLs
 * after a process restart without re-uploading.
 *
 * @param {string} scopeId
 * @returns {Promise<{ files: Array<object> }>}
 */
async function listScopeFiles(scopeId) {
  return getMonadix('/uploads/scopes/' + encodeURIComponent(scopeId) + '/files');
}

/**
 * Full delegation workflow (step-by-step — pauses at each boundary):
 *   Step 1 — Match:        call the match API, print ranked providers
 *   Step 2 — Confirm:      ask the user to pick a provider (or cancel)
 *   Step 3 — Publish & drive multi-turn conversation:
 *              3a reserve a conversation id (no credits)
 *              3b publish & wait for first turn (credits debited)
 *              3c loop on awaiting_consumer — collect user reply, send back
 *   Step 4 — Show Results: print the provider's output and usage summary
 *   Step 5 — Rate:         offer the user a 1–5 star rating prompt
 *
 * @param {string} description   - Task description (max 2000 chars)
 * @param {object} [inputData]   - Optional structured input for the provider
 */
async function delegateTask(description, inputData) {
  // Step 1 — Match: find providers (no credits spent)
  console.log('Matching providers...');
  var matchData = await callMonadix('/network/match', { description: description, limit: 5 });
  var matches = matchData.matches;
  if (!matches.length) {
    console.log('No providers available for this task.');
    return;
  }

  // --- PAUSE after Step 1: show results and wait for user selection ---
  console.log('\n[Step 1 complete — Match results]');
  console.log('\nFound ' + matches.length + ' provider(s):');
  matches.forEach(function (m, i) {
    var ratingStr = (m.provider.averageRating != null && m.provider.ratingCount > 0)
      ? ('\u2605 ' + Number(m.provider.averageRating).toFixed(1) + ' (' + m.provider.ratingCount + ' rating' + (m.provider.ratingCount === 1 ? '' : 's') + ')')
      : 'No ratings yet';
    console.log((i + 1) + '. ' + m.provider.name + ' (' + Math.round(m.score * 100) + '% match)  ' + ratingStr);
    console.log('   Capability: ' + m.capability.description);
  });

  var rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  // Step 2 — Confirm: wait for the user to choose a provider
  var answer = await prompt(rl, '\nWhich provider? (number or "cancel"): ');

  if (answer.trim().toLowerCase() === 'cancel') {
    rl.close();
    console.log('Delegation cancelled.');
    return;
  }
  var idx = parseInt(answer.trim(), 10) - 1;
  if (isNaN(idx) || idx < 0 || idx >= matches.length) {
    rl.close();
    console.log('Invalid selection.');
    return;
  }
  var chosen = matches[idx];

  // Step 3a — Reserve a conversation id (no credits, no provider contacted).
  // NEVER call this twice for the same logical request.
  console.log('\nReserving conversation id for ' + chosen.provider.name + '...');
  var draftPayload = {
    description: description,
    preMatchedProviderId: chosen.provider.id,
    preMatchedScore: chosen.score,
    preMatchedCapabilityDescription: chosen.capability.description,
  };
  if (inputData !== undefined) {
    draftPayload.input = inputData;
  }
  var draft = await callMonadix('/network/conversations/draft', draftPayload);
  var taskId = draft.task.id;
  console.log('Conversation reserved: ' + taskId + '. Publishing...');

  // Step 3b — Publish (fire-and-forget) then long-poll for the first
  // terminal/awaiting_consumer status.
  var ack = await publishConversation(taskId);
  if (!ack) {
    rl.close();
    console.log('\nDelegation failed (publish rejected).');
    return;
  }
  var snap = TERMINAL_STATUSES[ack.status] ? ack : await waitUntilTerminal(taskId);
  if (!snap) {
    rl.close();
    console.log('\nDelegation failed (status read error).');
    return;
  }

  // Step 3c — Reply loop. Continues until the conversation is terminal.
  while (snap.status === 'awaiting_consumer') {
    var prompt_ = snap.pendingPrompt || {};
    console.log('\n[Provider needs input]');
    console.log('Question: ' + (prompt_.question || '(no question text)'));
    if (prompt_.schema) {
      console.log('Expected shape: ' + JSON.stringify(prompt_.schema));
    }
    var userReply = await prompt(rl, 'Your reply (or "abort"): ');
    if (userReply.trim().toLowerCase() === 'abort') {
      console.log('Aborting conversation...');
      await closeConversation(taskId, 'User aborted from helper.');
      rl.close();
      return;
    }
    ack = await respondToProvider(taskId, userReply.trim(), genClientTurnId());
    if (!ack) {
      rl.close();
      console.log('\nDelegation failed (message rejected).');
      return;
    }
    snap = TERMINAL_STATUSES[ack.status] ? ack : await waitUntilTerminal(taskId);
    if (!snap) {
      rl.close();
      console.log('\nDelegation failed (status read error).');
      return;
    }
  }
  rl.close();

  // Step 4 — Show Results
  if (snap.status === 'completed') {
    console.log('\n[Step 4 — Results]');
    console.log('\nTask completed!');
    console.log('Output:', JSON.stringify(snap.result, null, 2));
    if (snap.creditCost != null) {
      console.log('Credits consumed: ' + snap.creditCost);
    }

    // Step 5 — Rate (completed only)
    var rl2 = readline.createInterface({ input: process.stdin, output: process.stdout });
    var providerName = chosen.provider.name;
    console.log('\n[Step 5 — Rate this provider]');
    console.log('How would you rate ' + providerName + '\'s work on this task? (1–5 stars)');
    console.log('\u2605 1 = Poor   \u2605 3 = Good   \u2605 5 = Excellent');
    var ratingAnswer = await prompt(rl2, 'Enter a number 1–5, or "skip": ');
    rl2.close();
    var rating = parseInt(String(ratingAnswer).trim(), 10);
    if (rating >= 1 && rating <= 5) {
      var err = await rateTask(taskId, rating);
      console.log(err ? 'Rating not recorded: ' + err : 'Rated ' + rating + '★');
    } else {
      console.log('Rating skipped.');
    }
  } else {
    // `failed` — server has already refunded any debited credits.
    var output = snap.output || {};
    var reason = output.error || output.reason || 'no reason given';
    console.log('\nTask failed: ' + reason + '. No credits charged.');
  }
}

module.exports = {
  callMonadix: callMonadix,
  getMonadix: getMonadix,
  delegateTask: delegateTask,
  publishConversation: publishConversation,
  respondToProvider: respondToProvider,
  closeConversation: closeConversation,
  rateTask: rateTask,
  createUploadScope: createUploadScope,
  uploadFile: uploadFile,
  listScopeFiles: listScopeFiles,
};

// Example invocation (runs when executed directly)
if (require.main === module) {
  delegateTask(
    'Summarise the key clauses of the attached contract and identify any unusual terms.',
    { contractText: '...' }
  ).catch(function (err) {
    console.error(err);
    process.exit(1);
  });
}

File: monadix-consumer/monadix.py
"""
monadix.py — Monadix consumer helper (Python 3.9+, stdlib only)

Place this file alongside monadix.key and monadix.signing-key.
Run:  python monadix.py
"""

import hashlib
import hmac
import json
import time
import urllib.error
import urllib.request
from pathlib import Path

BASE_DIR = Path(__file__).parent


def _load_key(filename: str) -> str:
    return (BASE_DIR / filename).read_text(encoding="utf-8").strip()


def _signed_request(method: str, path: str, body: dict | None = None) -> tuple[int, dict]:
    api_key = _load_key("monadix.key")
    signing_key = _load_key("monadix.signing-key")
    raw_body = "" if body is None else json.dumps(body, separators=(",", ":"))
    timestamp = str(int(time.time() * 1000))
    signed_payload = f"{timestamp}.{raw_body}"
    signature = hmac.new(
        signing_key.encode(), signed_payload.encode(), hashlib.sha256
    ).hexdigest()
    data = raw_body.encode() if raw_body else None
    headers = {
        "Authorization": f"Bearer {api_key}",
        "X-Monadix-Timestamp": timestamp,
        "X-Monadix-Signature": signature,
    }
    if method == "POST":
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(
        f"https://api.monadix.ai{path}",
        data=data,
        headers=headers,
        method=method,
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            text = resp.read().decode()
            return resp.status, (json.loads(text) if text else {})
    except urllib.error.HTTPError as err:
        text = err.read().decode() if err.fp else ""
        try:
            parsed = json.loads(text) if text else {}
        except json.JSONDecodeError:
            parsed = {"raw": text}
        raise MonadixAPIError(err.code, parsed, text) from err


class MonadixAPIError(Exception):
    def __init__(self, status: int, body: dict, text: str) -> None:
        super().__init__(f"Monadix API error: {status} {text}")
        self.status = status
        self.body = body


def call_monadix(path: str, body: dict | None = None) -> dict:
    """
    Send a signed POST request to the Monadix API.
    Credentials are read fresh from disk on every call — never cached.

    Args:
        path: API path, e.g. '/network/match'
        body: Request payload (will be JSON-serialised); None → empty body

    Returns:
        Parsed JSON response body as a dict

    Raises:
        MonadixAPIError: on 4xx/5xx responses (with .status and .body)
        urllib.error.URLError: on network errors
    """
    _, body_out = _signed_request("POST", path, body)
    return body_out


def get_monadix(path: str) -> dict:
    """Send a signed GET request. Empty body by definition."""
    _, body_out = _signed_request("GET", path, None)
    return body_out


def rate_task(task_id: str, rating: int) -> str | None:
    """
    Submit a 1-5 star rating for a previously completed task.
    Returns None on success or a string error message on a 4xx soft failure.
    """
    try:
        call_monadix(f"/network/tasks/{task_id}/rate", {"rating": rating})
        return None
    except (MonadixAPIError, urllib.error.URLError) as err:
        return str(err)


# ---------------------------------------------------------------------------
# Conversation primitives
# ---------------------------------------------------------------------------

import uuid

# Server-side per-call long-poll cap for `GET /network/tasks/<id>/status?wait=`.
# The server enforces a hard ceiling of 55_000 ms regardless.
STATUS_LONG_POLL_MS = 55_000

# Statuses at which the consumer must stop polling and either surface the
# result to the user or solicit a reply. Anything else means the provider is
# still working and the consumer should keep polling — the SERVER (not this
# helper) is responsible for the per-turn (10 min) and per-conversation (30
# min) wall-clock ceilings, so polling is bounded automatically.
TERMINAL_STATUSES = {"completed", "failed", "awaiting_consumer"}


def _gen_client_turn_id() -> str:
    return str(uuid.uuid4())


def _post_with_transport_retry(path: str, body: dict | None, *, label: str) -> dict | None:
    """POST and retry once on a transport/5xx error.

    The publish and messages endpoints are idempotent server-side (publish
    branches on persisted task status; messages dedupes on `clientTurnId`),
    so a fresh retry never double-bills or duplicates a turn. 4xx responses
    other than 409 are surfaced as terminal failures; 409 means the server
    cap was reached and is also terminal.
    """
    for attempt in (1, 2):
        try:
            return call_monadix(path, body)
        except MonadixAPIError as err:
            if err.status == 409:
                print(f"{label} rejected by server cap (409). Aborting. {err}")
                return None
            if 400 <= err.status < 500:
                print(f"{label} failed with {err.status} — not retryable. {err}")
                return None
            if attempt == 2:
                print(f"{label} failed after {attempt} attempts: {err}")
                return None
            print(f"{label} attempt {attempt} failed ({err.status}) — retrying once.")
        except urllib.error.URLError as err:
            if attempt == 2:
                print(f"{label} failed after {attempt} attempts: {err}")
                return None
            print(f"{label} attempt {attempt} failed (network) — retrying once.")
    return None


def wait_until_terminal(task_id: str) -> dict | None:
    """Long-poll `GET /network/tasks/<id>/status?wait=55000` until the task
    reaches `completed`, `failed`, or `awaiting_consumer`.

    The server enforces both the per-turn (10 min) and per-conversation
    (30 min) wall-clock ceilings; either being exceeded transitions the
    task to `failed` and refunds credits unconditionally. The consumer
    therefore NEVER needs to time out itself — every long-poll iteration
    is bounded by the server, and looping indefinitely is safe.

    Returns the final status snapshot, or None on a terminal API error.
    """
    while True:
        try:
            snap = get_monadix(
                f"/network/tasks/{task_id}/status?wait={STATUS_LONG_POLL_MS}"
            )
        except MonadixAPIError as err:
            print(f"Status read failed with {err.status}: {err}")
            return None
        except urllib.error.URLError as err:
            print(f"Status read network error: {err} — retrying.")
            continue
        if snap.get("status") in TERMINAL_STATUSES:
            return snap
        # pending / matched / executing — provider is still working; loop again.


def publish_conversation(task_id: str) -> dict | None:
    """Publish a previously-reserved conversation draft. Fire-and-forget.

    Returns the dispatch acknowledgment (`{task, status, ...}`) or None on
    terminal failure. The returned `status` may be `pending` / `matched` /
    `executing` (in flight — call `wait_until_terminal`), `awaiting_consumer`
    / `completed` (terminal — payload already includes `pendingPrompt` /
    `result`), or `failed` (terminal failure).
    """
    return _post_with_transport_retry(
        f"/network/conversations/{task_id}/publish", None, label="Publish"
    )


def respond_to_provider(task_id: str, content, client_turn_id: str | None = None) -> dict | None:
    """Send a single consumer follow-up message. Fire-and-forget.

    Always sends a `clientTurnId` (auto-generated when omitted) so a
    transport retry is dedupe-safe server-side.
    """
    turn_id = client_turn_id or _gen_client_turn_id()
    return _post_with_transport_retry(
        f"/network/conversations/{task_id}/messages",
        {"content": content, "clientTurnId": turn_id},
        label="Message",
    )


def close_conversation(task_id: str, reason: str | None = None) -> dict:
    """Idempotently close a conversation. Safe to call multiple times."""
    body = {"reason": reason} if reason else {}
    try:
        return call_monadix(f"/network/conversations/{task_id}/close", body)
    except (MonadixAPIError, urllib.error.URLError) as err:
        return {"error": str(err)}


# ---------------------------------------------------------------------------
# Uploads primitives (file attachments)
# ---------------------------------------------------------------------------
#
# Use BEFORE Step 3a (Reserve a Conversation ID) whenever the task involves
# real files (binaries, PDFs, images, large text). Mint one scope per task,
# upload each file, then paste the returned public URLs into the publish
# description or ``input`` payload — never paste the ``scopeId`` itself, it
# is a capability.

import mimetypes
import os


def create_upload_scope() -> dict:
    """Mint a fresh upload scope.

    Returns ``{"scopeId": "usc_...", "limits": {"maxFileSizeBytes": ...}}``.
    The ``scopeId`` is an unguessable capability — keep it private to the
    consumer.
    """
    return call_monadix("/uploads/scopes", {})


def upload_file(
    scope_id: str,
    local_file_path: str | os.PathLike,
    *,
    filename: str | None = None,
    content_type: str | None = None,
) -> dict:
    """Upload a single file to a scope using the pre-signed PUT flow.

    Two HTTP calls happen here:
      1. ``POST /uploads/scopes/<scopeId>/files/sign`` (Monadix-signed JSON)
         returns ``{uploadUrl, publicUrl, requiredHeaders, ...}``.
      2. ``PUT <uploadUrl>`` with the raw file bytes and the required
         ``Content-Type``. No Monadix credentials are sent on the PUT — the
         URL itself is the credential.

    Returns a flat dict ``{fileId, scopeId, storagePath, publicUrl,
    originalName, sizeBytes, contentType}``.
    """
    local_file_path = Path(local_file_path)
    file_bytes = local_file_path.read_bytes()
    upload_name = filename or local_file_path.name
    upload_type = content_type or (
        mimetypes.guess_type(upload_name)[0] or "application/octet-stream"
    )

    # Step 1: mint a pre-signed upload URL via the standard Monadix HMAC flow.
    signed = call_monadix(
        f"/uploads/scopes/{scope_id}/files/sign",
        {"filename": upload_name, "contentType": upload_type},
    )

    # Step 2: PUT the raw bytes directly to Supabase Storage. No Monadix headers.
    put_req = urllib.request.Request(
        signed["uploadUrl"],
        data=file_bytes,
        headers=signed["requiredHeaders"],
        method="PUT",
    )
    try:
        with urllib.request.urlopen(put_req, timeout=120) as resp:
            resp.read()  # drain Supabase receipt; we don't need it
    except urllib.error.HTTPError as err:
        text = err.read().decode() if err.fp else ""
        raise MonadixAPIError(err.code, {"raw": text}, text) from err

    return {
        "fileId": signed["fileId"],
        "scopeId": signed["scopeId"],
        "storagePath": signed["storagePath"],
        "publicUrl": signed["publicUrl"],
        "originalName": signed["originalName"],
        "sizeBytes": len(file_bytes),
        "contentType": upload_type,
    }


def list_scope_files(scope_id: str) -> dict:
    """List files already uploaded into a scope.

    Useful for recovering URLs after a process restart without re-uploading.
    """
    return get_monadix(f"/uploads/scopes/{scope_id}/files")


def delegate_task(description: str, input_data: dict | None = None) -> None:
    """
    Full delegation workflow (step-by-step — pauses at each boundary):
      Step 1 — Match:        call the match API, print ranked providers
      Step 2 — Confirm:      ask the user to pick a provider (or cancel)
      Step 3 — Publish & drive multi-turn conversation
      Step 4 — Show Results: print the provider's output and usage summary
      Step 5 — Rate:         offer the user a 1–5 star rating prompt

    Args:
        description: Task description (max 2000 chars)
        input_data:  Optional structured input for the provider
    """
    # Step 1 — Match: find providers (no credits spent)
    print("Matching providers...")
    data = call_monadix("/network/match", {"description": description, "limit": 5})
    matches = data.get("matches", [])
    if not matches:
        print("No providers available for this task.")
        return

    # --- PAUSE after Step 1: show results and wait for user selection ---
    print("\n[Step 1 complete — Match results]")
    print(f"\nFound {len(matches)} provider(s):")
    for i, m in enumerate(matches, 1):
        avg = m['provider'].get('averageRating')
        count = m['provider'].get('ratingCount', 0)
        if avg is not None and count > 0:
            rating_str = f"\u2605 {avg:.1f} ({count} rating{'s' if count != 1 else ''})"
        else:
            rating_str = "No ratings yet"
        print(f"{i}. {m['provider']['name']} ({round(m['score'] * 100)}% match)  {rating_str}")
        print(f"   Capability: {m['capability']['description']}")

    # Step 2 — Confirm: wait for the user to choose a provider
    answer = input("\nWhich provider? (number or 'cancel'): ").strip().lower()
    if answer == "cancel":
        print("Delegation cancelled.")
        return
    try:
        idx = int(answer) - 1
        assert 0 <= idx < len(matches)
    except (ValueError, AssertionError):
        print("Invalid selection.")
        return
    chosen = matches[idx]

    # Step 3a — Reserve a conversation id (no credits, no provider contacted).
    # NEVER call this twice for the same logical request.
    print(f"\nReserving conversation id for {chosen['provider']['name']}...")
    draft_payload: dict = {
        "description": description,
        "preMatchedProviderId": chosen["provider"]["id"],
        "preMatchedScore": chosen["score"],
        "preMatchedCapabilityDescription": chosen["capability"]["description"],
    }
    if input_data is not None:
        draft_payload["input"] = input_data
    draft = call_monadix("/network/conversations/draft", draft_payload)
    task_id = draft["task"]["id"]
    print(f"Conversation reserved: {task_id}. Publishing...")

    # Step 3b — Publish (fire-and-forget) then long-poll for the first
    # terminal/awaiting_consumer status.
    ack = publish_conversation(task_id)
    if ack is None:
        print("\nDelegation failed (publish rejected).")
        return
    snap = ack if ack.get("status") in TERMINAL_STATUSES else wait_until_terminal(task_id)
    if snap is None:
        print("\nDelegation failed (status read error).")
        return

    # Step 3c — Reply loop. Continues until the conversation is terminal.
    while snap.get("status") == "awaiting_consumer":
        pending = snap.get("pendingPrompt", {}) or {}
        print("\n[Provider needs input]")
        print(f"Question: {pending.get('question', '(no question text)')}")
        if pending.get("schema"):
            print(f"Expected shape: {json.dumps(pending['schema'])}")
        user_reply = input('Your reply (or "abort"): ').strip()
        if user_reply.lower() == "abort":
            print("Aborting conversation...")
            close_conversation(task_id, "User aborted from helper.")
            return
        ack = respond_to_provider(task_id, user_reply, _gen_client_turn_id())
        if ack is None:
            print("\nDelegation failed (message rejected).")
            return
        snap = ack if ack.get("status") in TERMINAL_STATUSES else wait_until_terminal(task_id)
        if snap is None:
            print("\nDelegation failed (status read error).")
            return

    # Step 4 — Show Results
    if snap.get("status") == "completed":
        print("\n[Step 4 — Results]")
        print("\nTask completed!")
        print("Output:", json.dumps(snap.get("result"), indent=2))
        credit_cost = snap.get("creditCost")
        if credit_cost is not None:
            print(f"Credits consumed: {credit_cost}")

        # Step 5 — Rate (completed only)
        provider_name = chosen["provider"]["name"]
        print(f"\n[Step 5 — Rate this provider]")
        print(f"How would you rate {provider_name}'s work on this task? (1–5 stars)")
        print("\u2605 1 = Poor   \u2605 3 = Good   \u2605 5 = Excellent")
        rating_answer = input('Enter a number 1–5, or "skip": ').strip()
        try:
            rating = int(rating_answer)
        except ValueError:
            rating = 0
        if 1 <= rating <= 5:
            err = rate_task(task_id, rating)
            print(f"Rating not recorded: {err}" if err else f"Rated {rating}\u2605")
        else:
            print("Rating skipped.")
    else:
        # `failed` — server has already refunded any debited credits.
        output = snap.get("output") or {}
        reason = output.get("error") or output.get("reason") or "no reason given"
        print(f"\nTask failed: {reason}. No credits charged.")


if __name__ == "__main__":
    delegate_task(
        "Summarise the key clauses of the attached contract and identify any unusual terms.",
        {"contractText": "..."},
    )
